include/range/v3/view/cache1.hpp

```diff
@@ -0,0 +1,173 @@
+/// \file
+// Range v3 library
+//
+//  Copyright Eric Niebler 2014-present
+//
+//  Use, modification and distribution is subject to the
+//  Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/ericniebler/range-v3
+//
+#ifndef RANGES_V3_VIEW_CACHE1_HPP
+#define RANGES_V3_VIEW_CACHE1_HPP
+
+#include <range/v3/range_fwd.hpp>
+
+#include <range/v3/detail/range_access.hpp>
+#include <range/v3/iterator/concepts.hpp>
+#include <range/v3/range/access.hpp>
+#include <range/v3/range/concepts.hpp>
+#include <range/v3/range/primitives.hpp>
+#include <range/v3/range/traits.hpp>
+#include <range/v3/utility/semiregular_box.hpp>
+#include <range/v3/view/adaptor.hpp>
+#include <range/v3/view/all.hpp>
+
+namespace ranges
+{
+    /// \addtogroup group-views
+    /// @{
+    template<typename Rng>
+    struct cache1_view : view_facade<cache1_view<Rng>, range_cardinality<Rng>::value>
+    {
+    private:
+        CPP_assert(input_range<Rng>);
+        friend range_access;
+        Rng rng_;
+        semiregular_box_t<range_value_t<Rng>> cache_;
+
+        struct cursor;
+
+        struct sentinel
+        {
+        private:
+            friend cursor;
+            sentinel_t<Rng> last_;
+
+        public:
+            sentinel() = default;
+            constexpr explicit sentinel(sentinel_t<Rng> last)
+              : last_(std::move(last))
+            {}
+        };
+
+        struct cursor
+        {
+        private:
+            cache1_view * parent_;
+            iterator_t<Rng> current_;
+
+        public:
+            using value_type = range_value_t<Rng>;
+            using single_pass = std::true_type;
+            using difference_type = range_difference_t<Rng>;
+
+            cursor() = default;
+
+            constexpr explicit cursor(cache1_view * parent,
+                                      iterator_t<Rng> current) noexcept
+              : parent_(parent)
+              , current_(std::move(current))
+            {}
+            range_value_t<Rng> && read() const noexcept
+            {
+                return static_cast<range_value_t<Rng> &&>(std::move(parent_->cache_));
+            }
+            void next()
+            {
+                if(++current_ != ranges::end(parent_->rng_))
+                    parent_->cache_ = *current_;
+            }
+            bool equal(cursor const & that) const
+            {
+                return current_ == that.current_;
+            }
+            bool equal(sentinel const & that) const
+            {
+                return current_ == that.last_;
+            }
+            CPP_member
+            auto distance_to(cursor const & that) const -> CPP_ret(difference_type)( //
```

Comment:
If this operation is used and it ever returns non-zero, it almost certainly indicates a bug in the calling code. I suggest dropping it.

Reply:
Returning a `cursor` from `end_cursor` makes this necessary. (I have vague oogie feelings about `common` single-pass ranges, but I'll drop this for now.)

Reply:
My thinking here is that _someday_ `std::vector`'s constructor will be hep to `sized_sentinel_for` and use it to reserve memory _even if the iterator is not forward_. Wouldn't that be nice?

Reply:
Oops - I suppose I should actually resolve the comment after saying "I'll drop this."